README.md
# styliteag/docker-keepalived-with-labels

This image uses https://github.com/osixia/docker-keepalived 
I adds helper script (notify.sh) which will update the Label of the host 

start-node-sh is a example how to start a keepalive on 3 nodes

The node where the keepalive is the MASTER, will have the KEEPALIVE-10.40.8.10=MASTER and KEEPALIVE-2001:DB8:4008:d0ce::10=MASTER

[hub]: https://hub.docker.com/r/sityliteag/docker-keepalived-with-labels

**A docker image to run Keepalived.**
> [keepalived.org](http://keepalived.org/)

This image require the kernel module ip_vs loaded on the host (`modprobe ip_vs`) and need to be run with : --cap-add=NET_ADMIN --net=host

    docker run --cap-add=NET_ADMIN --net=host -d styliteag/docker-keepalived-with-labels

## Beginner Guide

### Use your own Keepalived config
This image comes with a keepalived config file that can be easily customized via environment variables for a quick bootstrap,
but setting your own keepalived.conf is possible. 2 options:

- Link your config file at run time to `/container/service/keepalived/assets/keepalived.conf` :

      docker run --volume /data/my-keepalived.conf:/container/service/keepalived/assets/keepalived.conf --detach styliteag/docker-keepalived-with-labels

- Add your config file by extending or cloning this image, please refer to the [Advanced User Guide](#advanced-user-guide)

### Fix docker mounted file problems

You may have some problems with mounted files on some systems. The startup script try to make some file adjustment and fix files owner and permissions, this can result in multiple errors. See [Docker documentation](https://docs.docker.com/v1.4/userguide/dockervolumes/#mount-a-host-file-as-a-data-volume).

To fix that run the container with `--copy-service` argument :

		docker run [your options] styliteag/docker-keepalived-with-labels --copy-service

### Debug

The container default log level is **info**.
Available levels are: `none`, `error`, `warning`, `info`, `debug` and `trace`.

Example command to run the container in `debug` mode:

	docker run --detach styliteag/docker-keepalived-with-labels --loglevel debug

See all command line options:

	docker run styliteag/docker-keepalived-with-labels --help


## Environment Variables

Environment variables defaults are set in **image/environment/default.yaml**

See how to [set your own environment variables](#set-your-own-environment-variables)


- **KEEPALIVED_INTERFACE**: Keepalived network interface. Defaults to `eth0`
- **KEEPALIVED_PASSWORD**: Keepalived password. Defaults to `d0cker`
- **KEEPALIVED_PRIORITY** Keepalived node priority. Defaults to `150`
- **KEEPALIVED_ROUTER_ID** Keepalived virtual router ID. Defaults to `51`

- **KEEPALIVED_UNICAST_PEERS** Keepalived unicast peers. Defaults to :
      - 192.168.1.10
      - 192.168.1.11

  If you want to set this variable at docker run command add the tag `#PYTHON2BASH:` and convert the yaml in python:

      docker run --env KEEPALIVED_UNICAST_PEERS="#PYTHON2BASH:['192.168.1.10', '192.168.1.11']" --detach styliteag/docker-keepalived-with-labels

  To convert yaml to python online : http://yaml-online-parser.appspot.com/


- **KEEPALIVED_VIRTUAL_IPS** Keepalived virtual IPs. Defaults to :

      - 192.168.1.231
      - 192.168.1.232

  If you want to set this variable at docker run command convert the yaml in python, see above.

- **KEEPALIVED_NOTIFY** Script to execute when node state change. Defaults to `/container/service/keepalived/assets/notify.sh`

- **KEEPALIVED_COMMAND_LINE_ARGUMENTS** Keepalived command line arguments; Defaults to `--log-detail --dump-conf`

- **KEEPALIVED_STATE** The starting state of keepalived; it can either be MASTER or BACKUP.

### Set your own environment variables

#### Use command line argument
Environment variables can be set by adding the --env argument in the command line, for example:

    docker run --env KEEPALIVED_INTERFACE="eno1" --env KEEPALIVED_PASSWORD="password!" \
    --env KEEPALIVED_PRIORITY="100" --detach styliteag/docker-keepalived-with-labels


#### Link environment file

For example if your environment file is in :  /data/environment/my-env.yaml

	docker run --volume /data/environment/my-env.yaml:/container/environment/01-custom/env.yaml \
	--detach styliteag/docker-keepalived-with-labels

Take care to link your environment file to `/container/environment/XX-somedir` (with XX < 99 so they will be processed before default environment files) and not  directly to `/container/environment` because this directory contains predefined baseimage environment files to fix container environment (INITRD, LANG, LANGUAGE and LC_CTYPE).

## Advanced User Guide

### Under the hood: osixia/light-baseimage

This image is based on osixia/light-baseimage.
More info: https://github.com/osixia/docker-light-baseimage

## Changelog

Please refer to: [CHANGELOG.md](CHANGELOG.md)


CHANGELOG.md
# Changelog
This file only reflects the changes that are made in this project.
Please refer to the upstream [keepalived changelog](https://github.com/acassen/keepalived/blob/master/ChangeLog) for the list of changes in keepalived.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project follows keepalived versioning.

## master - 2019-05-05
### Initial
  - First Version


notify.sh
#!/bin/bash

# for ANY state transition.
           # "notify" script is called AFTER the notify_* script(s) and
           # is executed with 4 additional arguments after the configured
           # arguments provided by Keepalived:
           #   $(n-3) = "GROUP"|"INSTANCE"
           #   $(n-2) = name of the group or instance
           #   $(n-1) = target state of transition (stop only applies to instances)
           #            ("MASTER"|"BACKUP"|"FAULT"|"STOP")
           #   $(n)   = priority value
           #   $(n-3) and $(n-1) are ALWAYS sent in uppercase, and the possible
           #
           # strings sent are the same ones listed above
           #   ("GROUP"/"INSTANCE", "MASTER"/"BACKUP"/"FAULT"/"STOP")
           # (note: STOP is only applicable to instances)
           #           notify <STRING>|<QUOTED-STRING> [username [groupname]]

#echo $0 $@ >> /tmp/notify.log

if [ -x /usr/bin/docker -a -r /var/run/docker.sock ] ; then
	docker node update --label-add KEEPALIVE-$KEEPALIVED_VIRTUAL_IPS=$3 $HOSTNAME
	docker node update --label-add KEEPALIVE-$KEEPALIVED_ROUTER_ID=$3 $HOSTNAME
fi

start-node.sh
#!/bin/bash


HOSTNR=$(echo $HOSTNAME | tr -d "A-Za-z-")
PRIO=$[100 - $HOSTNR]
START=0

if [ "$HOSTNR" == "01" ] ; then
    PRIO=200
    START=1
    elif [ "$HOSTNR" == "02" ] ; then
    PRIO=150
    START=1
    elif [ "$HOSTNR" == "03" ] ; then
    PRIO=100
    START=1
fi

if [ "$START" == "0" ] ; then
    docker kill keepalived-ipv6
    docker kill keepalived-8
else
    modprobe ip_vs
    
    docker kill keepalived-8
    docker run --rm -d --name keepalived-8 \
    --cap-add=NET_ADMIN --net=host \
    -e KEEPALIVED_UNICAST_PEERS="#PYTHON2BASH:['10.40.8.11','10.40.8.12','10.40.8.13']" \
    -e KEEPALIVED_VIRTUAL_IPS=10.40.8.10 \
    -e KEEPALIVED_INTERFACE=ens192 \
    -e KEEPALIVED_PASSWORD=ens192 \
    -e KEEPALIVED_ROUTER_ID=8 \
    -e KEEPALIVED_PRIORITY=$PRIO \
    -v /var/run/docker.sock:/var/run/docker.sock \
    styliteag/docker-keepalived-with-labels:latest
    
    docker kill keepalived-ipv6
    docker run --rm -d --name keepalived-ipv6 \
    --cap-add=NET_ADMIN --net=host \
    -e KEEPALIVED_UNICAST_PEERS="#PYTHON2BASH:['2001:DB8:4008:d0ce::11','2001:DB8:4008:d0ce::12',2001:DB8:4008:d0ce::13']" \
    -e KEEPALIVED_VIRTUAL_IPS=2001:DB8:4008::10 \
    -e KEEPALIVED_INTERFACE=ens256 \
    -e KEEPALIVED_PASSWORD=ens256v6 \
    -e KEEPALIVED_ROUTER_ID=8 \
    -e KEEPALIVED_PRIORITY=$PRIO \
    -v /var/run/docker.sock:/var/run/docker.sock \
    styliteag/docker-keepalived-with-labels:latest
fi
